egs/wsj/s5/steps/nnet3/components.py

```diff
@@ -153,20 +173,32 @@ def AddLstmLayer(config_lines,
 
     # c1_t and c2_t defined below
     component_nodes.append("component-node name={0}_c_t component={0}_c input=Sum({0}_c1_t, {0}_c2_t)".format(name))
-    c_tminus1_descriptor = "IfDefined(Offset({0}_c_t, {1}))".format(name, lstm_delay)
+    if (state_preserving == "false"):
+        c_tminus1_descriptor = "IfDefined(Offset({0}_c_t, {1}))".format(name, lstm_delay)
+    else:
+        c_tminus1_descriptor = "Failover(Offset({0}_c_t, {1}), Offset(output_{0}_c_t_STATE_PREVIOUS_MINIBATCH, {1}))".format(name, lstm_delay)
 
     component_nodes.append("# i_t")
-    component_nodes.append("component-node name={0}_i1 component={0}_W_i-xr input=Append({1}, IfDefined(Offset({0}_{2}, {3})))".format(name, input_descriptor, recurrent_connection, lstm_delay))
+    if (state_preserving == "false"):
+        component_nodes.append("component-node name={0}_i1 component={0}_W_i-xr input=Append({1}, IfDefined(Offset({0}_{2}, {3})))".format(name, input_descriptor, recurrent_connection, lstm_delay))
+    else:
+        component_nodes.append("component-node name={0}_i1 component={0}_W_i-xr input=Append({1}, Failover(Offset({0}_{2}, {3}), Offset(output_{0}_{2}_STATE_PREVIOUS_MINIBATCH, {3})))".format(name, input_descriptor, recurrent_connection, lstm_delay))
     component_nodes.append("component-node name={0}_i2 component={0}_w_ic  input={1}".format(name, c_tminus1_descriptor))
     component_nodes.append("component-node name={0}_i_t component={0}_i input=Sum({0}_i1, {0}_i2)".format(name))
 
     component_nodes.append("# f_t")
-    component_nodes.append("component-node name={0}_f1 component={0}_W_f-xr input=Append({1}, IfDefined(Offset({0}_{2}, {3})))".format(name, input_descriptor, recurrent_connection, lstm_delay))
+    if (state_preserving == "false"):
+        component_nodes.append("component-node name={0}_f1 component={0}_W_f-xr input=Append({1}, IfDefined(Offset({0}_{2}, {3})))".format(name, input_descriptor, recurrent_connection, lstm_delay))
+    else:
+        component_nodes.append("component-node name={0}_f1 component={0}_W_f-xr input=Append({1}, Failover(Offset({0}_{2}, {3}), Offset(output_{0}_{2}_STATE_PREVIOUS_MINIBATCH, {3})))".format(name, input_descriptor, recurrent_connection, lstm_delay))
```

Comment:
In my implementation, I add to each example an additional input io with multiple time steps of the same variable (Ln 108-130 in nnet3-add-recurrent-io-to-egs.cc), just the same way as ordinary input. I think as long as the Index of those additional input io is correct (in terms of computability of all the component nodes after building the computation graph).